### PR TITLE
Extend motion prediction support

### LINKED ISSUE #38: MoP: Finish support across versions.
Ensure it works on all supported game versions.

### LINKED ISSUE #95: Update d2dx SGD2FreeRes
Can we get the updated version of SGD2FreeRes that fixes the cinematic cut scenes crashes in 1.09d

https://github.com/mir-diablo-ii-tools/SlashGaming-Diablo-II-Free-Resolution/issues/42

### LINKED ISSUE #99: Cutscenes are only 4:3 even in widescreen mode
Hi,

at first a BIG BIG thank you for creating this. It is simply amazing. I'm not sure if I need Resurrected anymore. 👍 

My only nitpick is that cutscenes are not scaled to fit a 16:9 screen. I'm used to this from Svens glide wrapper.
Would be cool if this feature gets implemented.

### LINKED ISSUE #105: Character movement will cause the ground item text to shake
Is this a bug?

### LINKED ISSUE #141: Framerate drop after unspecified time playing
I notice the framerate drops after a while. Solved by restarting the game.

Video example: https://clips.twitch.tv/ColdbloodedBenevolentWoodcockKreygasm-KRIYm4vTPSNcMFDT

### LINKED ISSUE #155: D2dx version not compatible
Good day!
I'm here to ask if its possible to have compatibility with d2 1.10 version ... I'm trying to play eastern sun but its not compatible... and i can't find a version 1.10f anywhere too =/

thank you very much in advance

### LINKED ISSUE #159: Actual mouse can enter black bar zone
When playing fullscreen or alt+enter windowed fullscreen with default resolution 4:3 and mouse lock active, the in-game cursor can't enter the black bar areas on the left and right, but the actual mouse can which results in lag time to move away from the black bars making the left and right sides of the screen slightly "sticky". Please add in an option to make the mouse lock prevent this from occurring. 

This happens in MedianXL, I'm assuming these options would make it happen for any version.

### LINKED ISSUE #165: Scroll of Inifuss has a black bar
https://imgur.com/a/oJQSsSf

Not important of course, might be a quick fix though.

### LINKED ISSUE #166: Fps drop after 20-30 mins of play
Hi, I’m using plugy and pd2 on season 4 with d2dx, and everything works great for the first 20 mins, and then the game seems to revert to 30 fps, and feels really jittery / choppy. I then restart the game and it works for 20 mins smooth as a whistle. I’m using windows 10 / and a 3080 card. I’ve tried all kinds of different configurations and it always ends with fps lag after 20 mins. It's like the background, effects, weather etc all continues running at 60fps, but the minions/AI and skill effects starts to run at 30fps. So I'm thinking this is something to do with predicted movements becoming out of sync. Is there anything else I could try? Is this a known problem with Pd2? Thanks so much, I’ll have to buy you a coffee.

### LINKED ISSUE #167: [1.13c] Function D2WIN_DrawText
Looks like you got the wrong address DrawText on version 1.13c.
```
D2FUNC(D2WIN, DrawText, void, __fastcall, (const wchar_t * wStr, int X, int Y, int nColor, int Centered), 0x12FA0)//D2Win.#10150
D2FUNC(D2WIN, DrawTextEx, void, __fastcall, (const wchar_t * wStr, int X, int Y, int nColor, int Centered, int TransLvl), 0x12F60)//D2Win.#10096
```
```
case D2Function::D2Win_DrawText:
			hModule = _hD2WinDll;
			ordinal = 10096;
			break;
```
to
```
case D2Function::D2Win_DrawText:
			hModule = _hD2WinDll;
			ordinal = 10150;
			break;
```

### LINKED ISSUE #172: d2dx 1.09d
Hello!

Can you answer whether You will continue to work on the modification? I encountered a problem related to the transition to the next act, when passing the current one (the game is crashing) with the "noresmod=true" (only with "true") and problem with hovering text.

Thank You for Your great work!

### PR DESCRIPTION
**Pull request by Jarcho**
Fixes #172
Fixes #166
Fixes #167
Fixes #165
Fixes #159
Fixes #155
Fixes #141
Fixes #99
Fixes #95
Fixes #38

Mostly fixes #105

Adds motion prediction to 1.10 and ports the motion prediction for text to all versions.

### CODE REVIEW

**Comment by Marcelo20XX** [comment]
I didn't test it on 1.09d yet but if the text works it will be amazing! Congratulations for a job well done @Jarcho ...

Now, since this project is mostly abandoned by its original author, could you please perhaps fix the bug with MoP disabling after playing for some time, its the only thing ruining the experience for everyone else right now and thanks again!!!

**Comment by Jarcho** [comment]
I haven't had that happen to me yet, so it's going to be a pain to debug. I know the motion prediction code has some n^2 parts to it which cause the frame rate to drop below 25fps on busy scenes which might be related.

Unit ID's are shortened to u16's instead of u32's which might also be the cause of the issue. Could try using the full u32 id and see if that ends up fixing it.

---

I'm working slowly on writing a different glide wrapper in my spare time. No idea when I'll have that done though.

**Comment by Jarcho** [comment]
Can you try [this](https://github.com/Jarcho/d2dx/releases/tag/test-motion-prediction) and see if it works? Note the builtin res mod is removed on this build. You'll need to provide your own version.

**Comment by Marcelo20XX** [comment]
I was back from work and I immediately saw you updated your repo, thanks my man!!! the belt and floating text issues are fixed. I am currently testing LoD 1.09d but I can be your beta tester for other patches too. Regarding the long session error, that's my main interest, I tried myself a couple of times to fix it but my knowledge of the D2 code isn't enough for it so I am finger crossing for a chance at a fix...

EDIT: I ran a test with the Sorceress on the entire ActV which was my previous field for testing the MoP issue and after having complete the entire act without the bug triggering I think its safe to assume the bug is fixed. The entire run took me 2 hours, where previously I triggered the bug after 20-30 mins playing this act...

**Comment by Jarcho** [comment]
That would mean the unit id's exceeding 16 bits was the issue. That doesn't fix the issue with frames dropping on busy scenes, but that needs a rewrite of the motion prediction code. Not something that will happen for a while.

If it ends up happening you can post back and I'll look into it some more.

**Comment by Marcelo20XX** [comment]
Exactly, I was having the FPS reverted back to 25FPS on busy scenes i.e. when fighting against Crushing Beast with an Imp mounted casting Inferno on Act V...

Thanks again for the fix and be sure I will report back if I have the issue happening again...

Btw, do you accept reporting other bug issues?

**Comment by Jarcho** [comment]
I can at least look in to things.

**Comment by jaydee2k** [comment]
> Can you try [this](https://github.com/Jarcho/d2dx/releases/tag/test-motion-prediction) and see if it works? Note the builtin res mod is removed on this build. You'll need to provide your own version.

Thanks a ton! Amazing work. Only thing your pre release build isn't working but when I compile from your source everything works perfectly with newest version of SGD2Freeres.

Your pre release build doesn't seem to recognize d2dx.cfg somehow...

**Comment by Jarcho** [comment]
The pre-release isn't quite the same as the source here. The source still uses the integrated res mod whereas the prerelease has it stripped out. You'd need to provide your own copy of `SGD2FreeRes.dll` alongside it.

**Comment by jaydee2k** [comment]
Yeah I know, I'm already using an external version of SGD2Freeres. Strangely only the compiled source is working for me. Works like a charm.

Pre release one ignores my d2dx.cfg and isn't using motion prediction. I have no idea why...

**Comment by Jarcho** [comment]
Strange. Ignoring `d2dx.cfg` would result in both being enabled so that's not the problem.

**Comment by CaptainRay1993** [comment]
> > Can you try [this](https://github.com/Jarcho/d2dx/releases/tag/test-motion-prediction) and see if it works? Note the builtin res mod is removed on this build. You'll need to provide your own version.
> 
> Thanks a ton! Amazing work. Only thing your pre release build isn't working but when I compile from your source everything works perfectly with newest version of SGD2Freeres.
> 
> Your pre release build doesn't seem to recognize d2dx.cfg somehow...

I tried to compile your d2dx-test-motion-prediction source in vs 2022 but got an error as followed:
Screenshot (47)
Any idea?

**Comment by CaptainRay1993** [comment]
> Can you try [this](https://github.com/Jarcho/d2dx/releases/tag/test-motion-prediction) and see if it works? Note the builtin res mod is removed on this build. You'll need to provide your own version.

I tried to compile your d2dx-test-motion-prediction source in vs 2022 but got an error as followed:

Screenshot (47)

Any idea?

**Comment by Marcelo20XX** [comment]
Just ignore that error, only thing it does is to copy the file glide3x.dll after compilation to a certain folder, the compiled file will be inside the Release folder. Also compile the source on "Release" mode not Debug...

**Comment by CaptainRay1993** [comment]
Thanks for your reply. I added float-point scaling factor from #133 and recompiled and it just works!

**Comment by jaydee2k** [comment]
I found a bug left with motion prediction for text in 1.09d. When you are walking and selecting npcs or containers like chests there will be still no motion prediction for these texts. Just a little annoyance.
In 1.13d theres motion prediction for these texts too.

**Comment by Marcelo20XX** [comment]
There is some polish needed for the hovering text MoP to work properly, it still has this bug: https://github.com/bolrog/d2dx/issues/105

The text sometimes jumps randomly in all versions and there is no MoP for Shrines too. Can you take a look at this @Jarcho thank you in advance...

**Comment by whipowill** [comment]
Can you guys provide a compiled download of glide3dx.dll for us?  Or give us instructions?  The file I found [here ](https://github.com/Jarcho/d2dx/releases/tag/test-motion-prediction)does not work in that it doesn't give me widescreen resolutions.

**Comment by Jarcho** [comment]
You'll need a copy of SGD2FreeRes for that version. This is the latest: https://github.com/mir-diablo-ii-tools/SlashGaming-Diablo-II-Free-Resolution/releases/tag/3.0.2.6. Both `SGD2FreeRes.dll` and `SGD2FreeRes.mpq` need to be placed alongside `glide3x.dll`.

**Comment by Marcelo20XX** [comment]
Also @Jarcho , I know this is not a bug but more of a feature, but since it can be a quick change to the source... Could you perhaps fix the Cinematics videos showing with letter boxes on both the horizontal and vertical axis? I mean since they are close to 16:9 shouldn't be full screen for people that plays on widescreen resolutions?

**Comment by tomasz1986** [comment]
Is the compiled `glide3x.dll` limited to a particular patch? I've tried using it with 1.13c/d and 1.14d and the game does launch with it, but it runs without any motion prediction at all. This is with or without `SGD2FreeRes.dll` and `SGD2FreeRes.mpq`. For the record, I only play using the default 800x600 resolution.

**Comment by CaptainRay1993** [comment]
> Can you guys provide a compiled download of glide3dx.dll for us? Or give us instructions? The file I found [here ](https://github.com/Jarcho/d2dx/releases/tag/test-motion-prediction)does not work in that it doesn't give me widescreen resolutions.

You can try [my build](https://github.com/CaptainRay1993/d2dx/releases) which updates built-in SGD2FreeRes to the latest version and includes changes from #177 and #133.

**Comment by whipowill** [comment]
> You'll need a copy of SGD2FreeRes for that version. This is the latest: https://github.com/mir-diablo-ii-tools/SlashGaming-Diablo-II-Free-Resolution/releases/tag/3.0.2.6. Both `SGD2FreeRes.dll` and `SGD2FreeRes.mpq` need to be placed alongside `glide3x.dll`.

I did get this to work but it seems to ignore my ``d2dx.cfg`` file?  So I'm getting widescreen resolutions but not smooth FPS.

**Comment by Marcelo20XX** [comment]
@whipowill You can unzip this to your root folder: https://github.com/bolrog/d2dx/files/9659733/MoP_1.13c.zip or try @CaptainRay1993 fork repository release until Jarcho gets a release. Either way the zip is preconfigured to Widescreen and Motion Prediction enabled...

**Comment by tomasz1986** [comment]
The @CaptainRay1993's fork works great for me, and it does respect `d2dx.cfg` too. It basically seems to behave exactly the same as the official `glide3x.dll`. I haven't played the game enough yet to test whether the fps drop bug has been fixed though.

**Comment by Jarcho** [comment]
> Also @Jarcho , I know this is not a bug but more of a feature, but since it can be a quick change to the source... Could you perhaps fix the Cinematics videos showing with letter boxes on both the horizontal and vertical axis? I mean since they are close to 16:9 shouldn't be full screen for people that plays on widescreen resolutions?

This is a small pain. The videos are displayed via screen-blitting and the two intro videos are 4x3.

**Comment by Marcelo20XX** [comment]
Can we change it so it crops the letterboxes, I did this for the Diablo 1 Source on the Devilution X project, but I didn't find where to do that here...

Also we can leave alone the two intro videos as they can be easily skipped via a simple modification on the D2Launch.dll (if anyone wants to know how, you can ask me)

Another possibility could be, show them as stretched and then overscan the top and bottom part of the image?

**Comment by Jarcho** [comment]
I think you misunderstood something. What d2dx gets for display is a 640x480 buffer per video frame with no direct way to distinguish which video is playing. This is for the intro videos and cinematics. Both intro videos are actually 640x480 videos. The cinematics are 640x292. Removing the black border on the cinematics itself is easy, but it requires a way to distinguish which video is currently playing

**Comment by Marcelo20XX** [comment]
Assume the video played is a cinematic, it could be implemented as a toggle if the user wants the vanilla behavior, because there is a way to disable the intro videos by patching the D2Launch.dll file...

Could you show me how to do this as a proof of concept? And thank you for addressing this issue...

**Comment by ghost** [comment]
Here the video and the main screen are stretched
[d2dx.zip](https://github.com/bolrog/d2dx/files/9793854/d2dx.zip)
Taken from ZYEL452-beta1-public

**Comment by Marcelo20XX** [comment]
@madbmax Thanks for the file but it is not what I was asking for, there is still letterboxes at the bottom and at the top, but surely is a step on the right direction, I think @Jarcho has a solution for this issue but he doesn't want to apply it to the intro videos too...

**Comment by ghost** [comment]
@Marcelo20XX 
> Also we can leave alone the two intro videos as they can be easily skipped via a simple modification on the D2Launch.dll (if anyone wants to know how, you can ask me)

Hello, please tell me how to do this on 1.14d?

**Comment by ChaosMarc** [comment]
> Hello, please tell me how to do this on 1.14d?

Hex edit your game.exe (e.g with [HxD](https://mh-nexus.de/en/hxd/)) and change offset 00035284 from 85C0 to EB7A

**Comment by ghost** [comment]
@ChaosMarc Thank you very much

**Comment by Jarcho** [comment]
New build here: https://github.com/Jarcho/d2dx/releases/tag/2022-10-19

This should actually be working this time. I left a local change disabling motion prediction last time.

New things:
* This includes the latest SGD2FreeRes version which has support for 1.10 and 1.12 as well as a fix for the black tile problem (hopefully).
* Cinematics should use the whole screen (the opening shown when launching the game for the first time will be displayed letterbox, but every other cinematic should work, and selecting any from the menu works).
* Switching aspect ratios in game now works properly.

Edit: alt-enter is now working (accidentally broke it)

**Comment by CaptainRay1993** [comment]
> New build here: https://github.com/Jarcho/d2dx/releases/tag/2022-10-19
> 
> This should actually be working this time. I left a local change disabling motion prediction last time.
> 
> New things:
> 
> * This includes the latest SGD2FreeRes version which has support for 1.10 and 1.12 as well as a fix for the black tile problem (hopefully).
> * Cinematics should use the whole screen (the opening shown when launching the game for the first time will be displayed letterbox, but every other cinematic should work, and selecting any from the menu works).
> * Switching aspect ratios in game now works properly.
> 
> Edit: alt-enter is now working (accidentally broke it)

Thanks for your update. I just updated [my build](https://github.com/CaptainRay1993/d2dx/releases) which includes latest version of sgd2res and your changes.

**Comment by Jarcho** [comment]
Just fixed the text motion prediction for 1.09 and 1.10. I hooked the wrong functions originally so only item labels had motion prediction. Build has been updated.

**Comment by CaptainRay1993** [comment]
> Just fixed the text motion prediction for 1.09 and 1.10. I hooked the wrong functions originally so only item labels had motion prediction. Build has been updated.

Thanks again for your quick fix. I just updated [my build](https://github.com/CaptainRay1993/d2dx/releases) as well.

**Comment by Marcelo20XX** [comment]
@Jarcho

**Comment by Marcelo20XX** [comment]
@Jarcho Awesome update! I tested the new fullscreen cinematics and it displayed wrong for me, don't know if anyone else had any problems with them. So I made a few corrections to the source code if you want to take a look:

RenderContextResources
RenderContext

After that they looked the way I always wanted:

https://user-images.githubusercontent.com/15281206/197113518-dcc1f382-f289-476d-bc27-65325ab5b0dd.mp4

Also, did you considered opening your own repository, mainly to open issues there so we can better keep track of bugs instead of only using this thread for everything, just a suggestion though, no pressure of any kind, you already done a lot for the project and for that I am forever grateful...

**Comment by Jarcho** [comment]
What was it displaying as before those changes?

**Comment by Marcelo20XX** [comment]
Testing using LoD 1.09d on Windows 10...
The cinematic shows wrongly cropped:

https://user-images.githubusercontent.com/15281206/197123214-1bfd5027-1a33-46aa-bcb3-d8e26848eb2c.mp4

I had my desktop set at 1920x1080p

**Comment by ghost** [comment]
On LoD 1.14d on Windows 10
desktop set at 1280x720p
game size=[1068,600]
Same problem

**Comment by Jarcho** [comment]
I see that now. Definitely wrong.

**Comment by Jarcho** [comment]
Fixed. New [build](https://github.com/Jarcho/d2dx/releases/tag/2022-10-21) up.

Feel free to open issues on the fork. I'll get alerts either in this thread or over there.

**Comment by Marcelo20XX** [comment]
Thanks for the cinematics fix, now is working as intended. 

> Feel free to open issues on the fork. I'll get alerts either in this thread or over there.

That's the thing, I don't see any Issues tab to report bugs on your fork...

**Comment by Jarcho** [comment]
> That's the thing, I don't see any Issues tab to report bugs on your fork...

Fixed

**Comment by AngelOfLlife** [comment]
Hello. Is it possible to do main menu is 16:9? For example through "not keep aspect ratio"

**Comment by Jarcho** [comment]
The game runs the menu only at 800x600 (and I think it has since v1.0). All the artwork for the menu's are also made for 800x600. 

At a minimum, this would require artwork to put on the sides of the menu. This would still leave a 4x3 menu with the cursor boxed in, but would give something other than a black screen to look at. To actually get a 16x9 menu would require a fair bit of code editing. Unlike the in-game code, the menu code has no resolution switching support.

**Comment by AngelOfLlife** [comment]
Not real 16:9, but just stretching on the sides "the screen will be stretched without preserving the aspect ratio"
As here:

https://user-images.githubusercontent.com/116603991/197755653-bf6af372-cdbe-479e-badb-330f9c2147db.mp4

**Comment by CaptainRay1993** [comment]
I just updated [My build](https://github.com/CaptainRay1993/d2dx/releases/tag/0.99.529c) which includes the latest version of SGD2FreeRes and all changes from this PR.

**Comment by tomasz1986** [comment]
Recently, I've been testing the game with the patched `glide3x.dll` v0.99.529c from https://github.com/CaptainRay1993/d2dx/releases, but unfortunately, after leaving the game minimised for 1-2 days and then coming back to it, the screen started to shake a little when moving. I can't directly compare it with the original DLL from this repo to say whether there is any difference between the two, but still, the shaking makes the game impossible to play (e.g. names of items dropped on the ground keep moving left and right slightly) and a full restart is required to make everything smooth again. This is on Diablo II v1.13d, the original 800x600 resolution, and of course motion prediction turned on.

**Comment by Jarcho** [comment]
Is it just the text labels shaking?

---

@AngelOfLlife I can add a config option to ignore the aspect ratio everywhere.

**Comment by tomasz1986** [comment]
> Is it just the text labels shaking?

Unfortunately, it's everything, but with the text labels the shaking is just the most noticeable. I will try to record a video next time this happens (which I can probably reproduce simply by leaving the game minimised for a longer period of time).

**Comment by AngelOfLlife** [comment]
@Jarcho Yes. add please.

**Comment by Jarcho** [comment]
[New build up](https://github.com/Jarcho/d2dx/releases/tag/2022-11-02).

Adds a new option `nokeepaspectratio`. Can be set from either the config file or the command line as `-dxnokeepaspectratio`

**Comment by AngelOfLlife** [comment]
@Jarcho Excellent, thank you!

**Comment by tomasz1986** [comment]
> Is it just the text labels shaking?

I've reproduced the shaking again by leaving the game minimised for some time. Here is the video.

https://www.youtube.com/watch?v=WekBKio5v4I

If you look closely, you will see that the whole terrain seems to shake slightly when the character is moving, and of course the text labels keep changing their position ever so slightly. I limited the FPS for the recording to 60, but when I play normally, I've got the limit set at 160+ (on a 165 hz monitor), which only makes the problem even worse.

**Comment by zoran8844** [comment]
> > Is it just the text labels shaking?
> 
> I've reproduced the shaking again by leaving the game minimised for some time. Here is the video.
> 
> https://www.youtube.com/watch?v=WekBKio5v4I
> 
> If you look closely, you will see that the whole terrain seems to shake slightly when the character is moving, and of course the text labels keep changing their position ever so slightly. I limited the FPS for the recording to 60, but when I play normally, I've got the limit set at 160+ (on a 165 hz monitor), which only makes the problem even worse.

I had similar issue like yours couple of months back when I was playing D2 with D2DX. My FPS would fluctuate between 144 and 100 on the random time intervals (even in the town with char standing still).

I fixed it by making the real-time process exceptions of all the D2 .exe files ("Diablo II.exe" and "Game.exe") in my antivirus software along with D2 save folder "User\Saved Games\Diablo II\". If you are using PlugY you can try including "PlugY.exe" in the antivirus real-time process exception as well.

Hope this helps and sorry for out of nowhere replay :)

**Comment by Jarcho** [comment]
Looks like a slight desync with the game's frame rate. Currently testing a rewrite of the unit motion prediction code to see if that fixes the issue.

**Comment by Jarcho** [comment]
New [build](https://github.com/Jarcho/d2dx/releases/tag/2022-11-23) up. This should fix the issue.

**Comment by jaydee2k** [comment]
I don't know why but after your newest changes the game isn't fluid to me anymore. A bit choppy as if motion prediction isn't working at 100% anymore. Compiled my latest build before that at november 8 and this plays perfectly. Using Gsync with 155 hz/fps. Compiled myself and tried the release, both have constant microstutters for me. Feels like Diablo with 25fps again.

**Comment by tomasz1986** [comment]
I can confirm that everything does seem less smooth in comparison to the previous version. It's very noticeable, e.g. when you move the character around and look at the terrain. I play with FreeSync and the FPS capped at 162 on a 165 hz monitor.

**Comment by Jarcho** [comment]
Can you use [this version](https://github.com/Jarcho/d2dx/releases/download/2022-11-23/d2dx-2022-11-23-profile.zip) and post the `d2dx_log.txt` after. This will have some basic profiling info.

I'm guessing the problem is an unstable frame rate. D2DX can barely hit 100fps for me with some very unstable frame timings.

**Comment by tomasz1986** [comment]
Here you are.

[d2dx_log.txt](https://github.com/bolrog/d2dx/files/10085501/d2dx_log.txt)

**Comment by jaydee2k** [comment]
No this can't be the issue, I'm hitting stable 155 fps all the time... Doesn't matter which version.
Checked with MSI Afterburner...

[d2dx_log.txt](https://github.com/bolrog/d2dx/files/10085533/d2dx_log.txt)

**Comment by Jarcho** [comment]
[New version](https://github.com/Jarcho/d2dx/releases/tag/2022-11-24). Hopefully this fixes the issue.

**Comment by tomasz1986** [comment]
Unfortunately, the newest version still feels slightly "jittery" in comparison to the old ones 🙁. Also, the whole screen seems to shake a bit back and forth when the character stops moving. This can be seen immediately if you look at things like buildings in town, etc.

**Comment by Jarcho** [comment]
[Second attempt](https://github.com/Jarcho/d2dx/releases/tag/2022-11-25). This will get rid of the bounce back when the player stops moving. That's a trade-off between rendering the unit at the correct place, and bouncing the unit back when the prediction is wrong. For the player it's probably better to avoid bouncing as it affects the whole screen, but I've left the bounce on other units.

Hopefully the jitter is better this time around.

**Comment by tomasz1986** [comment]
Hmm, I've tested the newest release, and the bouncing seems to be gone, which is great, but honestly, it still feels weird when compared to the previous releases (e.g. d2dx-2022-11-02). For example, in the old version, when the character is running, I can keep looking at the terrain that moves with it, and despite the character changing position, the terrain still appears smooth and rather stable. However, in the newest versions, when I try to to do the same, the terrain appears blurry and more unstable. I'm not sure how to describe the effect better, but I would say that it resembles how the original D2 behaves stuck at 25 FPS and without motion prediction.

In other words, I think it's basically the same issue as described by @jaydee2k in https://github.com/bolrog/d2dx/pull/177#issuecomment-1326085269 above.

**Comment by Jarcho** [comment]
[Another attempt](https://github.com/Jarcho/d2dx/releases/tag/2022-11-28). If this doesn't fix it can you upload another copy of the log? The new version has a little bit more info dumped to it.

**Comment by tomasz1986** [comment]
Apologies for taking so long, but I've only now had time to test the newest version. Unfortunately, the background (terrain, etc.) still feels jittery and blurry when moving the character. Also, in comparison to the previous release, character movement changes seem super sharp.

[d2dx_log.txt](https://github.com/bolrog/d2dx/files/10146400/d2dx_log.txt)

**Comment by Jarcho** [comment]
[Yet another attempt](https://github.com/Jarcho/d2dx/releases/tag/2022-12-05). This should remove the hiccup every that was occurring every time the game updated positions. That was the only jitter I saw in the logs. This also switches to only predicting half a frame ahead which will help with any jerkiness caused by mispredicted positions.

Like last time if it still isn't smooth can you upload the log.

**Comment by Jarcho** [comment]
[New build](https://github.com/Jarcho/d2dx/releases/tag/2022-12-07)

**Comment by tomasz1986** [comment]
I've just tried the new build out. It does feel smoother than the last one. However, the terrain still feels unstable when the character is moving, resulting in it looking blurry. I can observe this, e.g. when looking straight at the terrain in front of the moving character.

[d2dx_log.txt](https://github.com/bolrog/d2dx/files/10185666/d2dx_log.txt)

For the record, the last version that doesn't display this kind of effect is https://github.com/Jarcho/d2dx/releases/tag/2022-11-02 (and, of course, the official releases from this repository). There, I can look at the terrain in the same situation and it remains stable all the time, even when the character is running quickly.

**Comment by Jarcho** [comment]
[New build](https://github.com/Jarcho/d2dx/releases/tag/2022-12-10). I don't know if this will fix blurriness or not. The change to the motion prediction is pretty small this time.

**Comment by tomasz1986** [comment]
I've tried out the new build but I can't really feel any difference between it and the last one.

[d2dx_log.txt](https://github.com/bolrog/d2dx/files/10207764/d2dx_log.txt)

**Comment by Jarcho** [comment]
[New build](https://github.com/Jarcho/d2dx/releases/tag/2022-12-18).

Were you using the new upscale method? It would be the default without a config file.

**Comment by Jarcho** [comment]
[New build](https://github.com/Jarcho/d2dx/releases/tag/2022-12-23).

This doesn't fully fix the jiggling text labels, but it should be pretty close.

**Comment by Jarcho** [comment]
[New build](https://github.com/Jarcho/d2dx/releases/tag/2022-12-24).

This only really has an effect at lower frame rates.

**Comment by tomasz1986** [comment]
Unfortunately, the situation seems to have got even worse now 🙁. The same blurry/unstable background when running is still present, and on top of that the game has started to visibly stutter every few seconds (even though the FPS remains at stable 162 all the time). I've tested both with my old config file and without any (i.e. using the default settings) but there was no difference.

Here is the log file, although it appears extremely small compared to the previous ones.

[d2dx_log.txt](https://github.com/bolrog/d2dx/files/10299863/d2dx_log.txt)

**Comment by Jarcho** [comment]
Just added the [profiled version](https://github.com/Jarcho/d2dx/releases/download/2022-12-24/d2dx-2022-12-24-profile.zip). Can you run that post the log?

**Comment by tomasz1986** [comment]
Here you go, although for some reason the game started in 640x480 resolution, which I then changed to 800x600. Also, I haven't observed the stuttering this time… No difference regarding background still being blurry while running.

[d2dx_log.txt](https://github.com/bolrog/d2dx/files/10300342/d2dx_log.txt)

Just a side question, as this isn't strictly related just to the PR itself, but I'd like to try doing some testing myself to find out which exact change is responsible for the difference/regression regarding terrain blurriness. What is the easiest way to compile the whole thing? Unfortunately, there don't seem to exist specific instructions anywhere. I'm more comfortable with building from the command line if that's of any significance.

**Comment by eleriaqueen** [comment]
Hello, I just tested the latest build (d2dx-2022-12-24) and I too can see some slight blurriness on the BG when moving (with Scaling Mode 3) but it's so slight it doesn't bother me at all, but it looks as if the area close to the character shows some pixelation as if "unstable" when moving though, not sure how to describe it.

On another note, would be cool if you (Jarcho) could fix the high CPU usage when window is minimised / tabbed-out from.
Thank you for your hard work.

Edit : Tabbing out during startup cinematics seems to affect the predictor in a weird way (Log Included) : 
```Using DXGI_FORMAT_R10G10B10A2_UNORM for the render buffer.
Sizes: desktop 1920x1080, window 800x600, game 640x480, render 800x600
MotionPredictor: Expecting frame update
MotionPredictor: Predict at 1111/2621 (+-2147483648) of a frame
MotionPredictor: Predict at -469/2621 (+-2147483648) of a frame
```
[d2dx_log.zip](https://github.com/bolrog/d2dx/files/10307687/d2dx_log.zip)

**Comment by Marcelo20XX** [comment]
> Here you go, although for some reason the game started in 640x480 resolution, which I then changed to 800x600. Also, I haven't observed the stuttering this time… No difference regarding background still being blurry while running.
> 
> [d2dx_log.txt](https://github.com/bolrog/d2dx/files/10300342/d2dx_log.txt)
> 
> Just a side question, as this isn't strictly related just to the PR itself, but I'd like to try doing some testing myself to find out which exact change is responsible for the difference/regression regarding terrain blurriness. What is the easiest way to compile the whole thing? Unfortunately, there don't seem to exist specific instructions anywhere. I'm more comfortable with building from the command line if that's of any significance.

The easiest way to compile the project is using Visual Studio and just choose release mode...

**Comment by Jarcho** [comment]
[d2dx-test.zip](https://github.com/bolrog/d2dx/files/10318329/d2dx-test.zip)

This partially reverts back to the older frame time tracking and removes the motion prediction. Hopefully this one ends up smooth.

**Comment by bolrog** [comment]
Very cool! I'd love to merge this, is it stable? I noted you had continued working on your branch.

I just don't have the time to work on this project anymore, so I'd be happy if someone carried it on.

**Comment by Jarcho** [comment]
The branch currently is not. I was working under the assumption you weren't coming back to this.

I can roll it back to something stable some time this weekend.

**Comment by Jarcho** [comment]
@bolrog Split off a few of the changes into different PR's. This was done mostly by cherry picking, but I haven't tested them. Hopefully they all compile and work.

**Comment by Jarcho** [comment]
[New release](https://github.com/Jarcho/d2dx/releases/tag/v0.100.0)

Unlike the last few build this uses the old motion prediction code. This still includes all the other bug fixes and improvements.
